### ISSUE TITLE
Add the activity type to the exported GPX files

### ISSUE THREAD

**Issue by GrazianoCapelli** [BasicAirData/GPSLogger#163]
As @TVEGit suggested into the issue #152, we could add the actiity type to exported GPX files.
Basing on what is written into the [GPX 1.1 Manual](https://www.topografix.com/gpx_manual.asp#keywords):
> ```<keywords>```
> Optional in: [```<gpx>```](https://www.topografix.com/gpx_manual.asp#gpx)
> ```<keywords>Hiking, NH, Presidential Range</keywords>```
> Keywords for indexing the GPX file with search engines. Comma separated.

the `<keywords>` tag could be a good place where to put the activity type.
Reading the example, the best approach to follow the guidelines is the following:
```xml
<keywords>walking</keywords>
```
Obviously the activity type should always be expressed in English, in order to keep the keywords homogeneous across the languages, thus we should hardcode the strings instead of use a string resource.
Since we can freely choose a convention, we could copy the activity definitions from Garmin:
- other
- mountaineering
- walking
- running
- cycling
- driving_general
- flying

Both GPX standards have the `<keywords>` field, but in different places (GPX 1.1 requires to place it into `<metadata>` section).
Here below two examples:

### GPX 1.0
```xml
<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.0"
     creator="BasicAirData GPS Logger 3.1.3-dev.20220205"
     xmlns="http://www.topografix.com/GPX/1/0"
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:schemaLocation="http://www.topografix.com/GPX/1/0 http://www.topografix.com/GPX/1/0/gpx.xsd">
<name>GPS Logger 20220205-173623</name>
<desc>First small outdoor activity with Iris</desc>
<time>2022-02-06T13:18:36Z</time>
<keywords>walking</keywords>
...
```
### GPX 1.1
```xml
<?xml version="1.0" encoding="UTF-8"?>
<gpx version="1.1"
     creator="BasicAirData GPS Logger 3.1.3-dev.20220205"
     xmlns="http://www.topografix.com/GPX/1/1"
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd">
<metadata> 
 <name>GPS Logger 20220205-173623</name>
 <desc>First small outdoor activity with Iris</desc>
 <time>2022-02-06T20:23:26Z</time>
 <keywords>walking</keywords>
</metadata>
...
```
Both the examples here above passed the Xerces validation test, and could be implemented.
We are used to validate our GPX files using Xerces (mine is the v3.2.3 for Linux):

```bash
SAXCount -v=always -n -s -f file_to_validate.gpx
```

**Comment by TVEgit** [comment]
I agree with the proposal, yet I'd like to say that I'd rather have GPS activities mapped to a custom list if keywords.
I don't see how conforming to Garmin types would be good for anyone but English speaking users.
As a matter of fact, what is the use case if this feature?
1. Making tracks searchable on the web?
2. Storing the information for archiving?
3. Passing information for postprocessing in a web hosting system? 
As far as I'm concerned, mainly 3), with 2) to some extent.
Garmin keywords would be good for 2), but only good for 1) for English speaking users. For 3), it could work but with likely obligation for the target system to do the translation. Without any syntax like the "Activity=" that I proposed (only comma separated keywords), that postprocessing would be harder since other keywords non referring to activities might be present in the file.
My proposal "Activity=keyword" would therefore tick all boxes, doing the job for all 3 use cases.

**Comment by GrazianoCapelli** [comment]
In general conforming means having a larger compatibility with the rest of the world, and this is the reason why we consider this aspect very important. In this case, the best choice seemed to be to follow the Garmin types, that is a standard-de-facto.
At worst, adopting an existing standard is still better than inventing another one anyway.

_I freely write what I thought:_

The use of the `Activity=...` seemed to me a bit out of <Keywords> use, thus I proposed a more standard approach (at least for XML documentation). Regarding this point, if you have some good examples of a similar use of `<Keywords>`, we are free to change our mind and adopt it.
I proposed to remove the `Activity=` because I thought to use homogeneous types, I understand that is impossible to use my approach with localized types.

Speaking of localized keywords, before to write my previous post I made some tests with Garmin Connect, and I verified that the type of activity is alwaye stored/exported in English language. I agree with their choice because:

1) Homogeneous keywords could be understood and easily managed by any software: one word has one meaning. It is more suitable for postprocessing in a web hosting system (as you rightly wrote), but it lends itself to better informative use also for archiving and aggregate research on the web (they would be listed when the users search one of the often researched Garmin keywords. It has limits, but is quite functional);
3) Localized keywords would need a translation in each language (with the foresight never to be modified later) and, as open and free software, we cannot effort and guarantee this.
Furthermore, as you seen on issue #152, we have in plan to add more activity types: 30 types added = 30 strings to translate in each language.
5) Localized keywords change when the user changes the language of the system (the app follows the system language), or when a language is removed because not yet maintained (like the Japanese in v3.1.0); this could be a small problem, but it would lead to aggregation issues on external platforms;

To sum up, we could add something like `Activity=...`, but unfortunately I think that we must exclude the localized types.

**Comment by TVEgit** [comment]
Good and clear points.
I haven't made any research to determine if the Garmin activities' syntax is being used and understood by other systems than theirs. If true, I am definitely convinced it's better to follow suit. Scale is of the essence for standards, de facto or official ones. However if the only reason is to do like them when I don't see any reason to use Garmin Connect for GPS Logger users.
The suggestion to add "Activity=" is only to add context for more robust postprocessing. It isn't really required if the added robustness is only theoretical. So I am not insisting on this request.

**Comment by GrazianoCapelli** [comment]
> This is why I am of the opinion that a user editable list would be best: 1) to put what you want and 2) to remove what you don't want.

I agree, the possibility to edit and create activity types (that should not be restricted) could be a step forward (that could be implemented later), and from another point of view could leave to the user burden and honor of translation.

The activity type has been initially implemented only to give to the user an extra visual identifying mark in order to help to find the track into the Tracklist (because we chosen to not integrate any map), but we will consider the good idea to enforce and customize the categorization for future implementations.
The `<keyword>`, at this point, would follow the definition of the user.

Anyway, at the moment we can only add the standard keywords, as written in my first post of the issue.
It's a start; I hope it will help you.

**Comment by TVEgit** [comment]
> Anyway, at the moment we can only add the standard keywords, as written in my first post of the issue.
> It's a start; I hope it will help you.

Yes, indeed.

Thank you.

**Comment by GrazianoCapelli** [comment]
@WRPSoft - Just to inform you that on commit 9a6a705 we added the optional `<bounds>` field to the `<gpx>` tag. If you want you might use it to center the map when you receive a GPX file via intent instead of starting from Frankfurt am Main.
We added this line:
```XML
<bounds minlat="48.89641204" minlon="13.23721145" maxlat="48.92826998" maxlon="13.25859562" />
```

**Comment by WRPSoft** [comment]
Dear @GrazianoCapelli,

Thanks for the information. I'll take a closer look.

At the moment I'm following this approach with WRPElevationChart for all kind of imports (for GPX sharing too): the map will always be initially centered on the whole track. 
This means that if BasicAirData GPSLogger transfers GPS coordinates (e.g. a recorded track), which should actually always be the case, the map will be centered accordingly. That actually works very well. To be honest, OSMDroid library (https://github.com/osmdroid/osmdroid) is doing the centering stuff 😊)

I use GPSLogger very often because I have to keep a driver's logbook sometimes and my app serves as external viewer. This is also the reason why I recently implemented an optional SPD curve. Therefore I have a good sense for this use case.

But again, I will take a look and eventually implement these bounds coordinates as a kind of fallback. 

PS: Thanks for the ongoing development of this awesome GPSLogger project 👍
